### DISCUSSION TITLE
Is it possible to create on each folder a `Readme.md` file?

### DISCUSSION THREAD

**Discussion by The-Lum** [plantuml/plantuml-stdlib#102] [Polls]
💡 Here is a poll for `Readme.md` creation.

**Comment by Potherca** [comment]
I think the question is, what should be in the file? @The-Lum If you have some idea about that, I could set up a first-draft MR to discuss things further.

**Reply by The-Lum** [reply]
FYI, here is a first PR:
- #113

**Comment by Potherca** [comment]
As a consumer of the `INFO` files (for instance in [plantuml-stdlib/update-stdlib.sh](https://github.com/plantuml-stdlib/update-stdlib.sh)), I definitely prefer YAML frontmatter over MD tables.

On the topic what should be in the file, I am still unsure. The `README.md` already in the projects themselves would not translate well into plantuml/plantuml-stdlib.

So what information should be present in the README and where should it come from?

**Comment by kirchsth** [comment]
Based on the last changes we have the details in the readme.md, but in the code itself I cannot query it (like `%version()` for PlantUML).
Shouldn't we add calls like `%libversion("C4")`, `%libversions()`, `%libversionjson("C4")` and `%libversionsjson()`?
And the json version contains only filled fields. 

Corresponding samples see below:   

[![https://www.plantuml.com/plantuml/uml/pLNVQzim47xNNt4pAqqXSTgGXn4ejEHfi4BPElOmob1i9T4gMqQIwxKb__i-HJPHLo-bbFsu1rUwq-bEttTVSsOTDwuk5EDAwQQGFsIvOkoGHgCHleSbByICvbEggqgOZ5j10oLl1BcreALMkJ17EDSTmpANjbByxbld8YKNXfix6iodeql5I_VvueFFoONDZAoSrEMChjOYFYjAlb324YShYqe9QlWTwMMRiq2FnCkSCbqKTIapxWHPlNGDDuAuoTRIYSpL1ZSqqgrnbcJfQkdahI2qiAhvIgJeWTTkhSsCiBmLncvRcQGd9-anF4OeWGVxA1hBHEdN5--li4UrscJOhfshx6myNg5ClKZHqBX20O_eoBeSxnb7pqQYDd8bKQM3s2_I5BtiF04vtjWdQ9v_-TH1r0DW70OqfPNMMO_QUS7lTRaDI-yVadHK6NqhSKLm-MEkqQYtN0eZIWVWxuMDCPglZRHqoQq5oXYM6BEJT-dn99rCzg2s7xH2AZqgP8bvMco_OtIx1pTsYC6ebIT-i42vdF30fLGAgu-2vuL8YnmEeNABONlpTZvjcpqz04JjUf3CfyaHYw9qcWJ-4nQoljl7ATZ7QTWYXnuOKU8vIcX6FZhqZfYyDj2lXd0yudDx-bze8dGItXgw2704Vyjd0ZSOJORadO5LxtkP61AssMACGU-b1dw07CT1JZTeIPlPIq-K6Wxiw9cH3_RGWq0adwU48Fw791Pg1PNvCh_gx34bCzfXqCl9VahsIorcHQcsf8tnjGFW6WSn_5zHaXwsOx-RbysmzNguuwPfmwwfA-7PUNo_ztGri0u5ADmLA6m5vdqmaEIuLBx6G_3qA17sOZd2_YTDmdulJ5Y_EhszIgBaRUF_T8J6EdgUXLXmiETLv-tlIW_slFvW_IBSWty8l5QEWG5YWOThnjX1smPaillXOj7_ISQySWWA0bwMOhb2U7lcdgIbKZjgjBdnipjeoI9FrZPIQMibHlw8GWrSrVuATaKR5CGdTYaLHWkHJjNiJ9GvXkud](https://www.plantuml.com/plantuml/svg/pLNVQzim47xNNt4pAqqXSTgGXn4ejEHfi4BPElOmob1i9T4gMqQIwxKb__i-HJPHLo-bbFsu1rUwq-bEttTVSsOTDwuk5EDAwQQGFsIvOkoGHgCHleSbByICvbEggqgOZ5j10oLl1BcreALMkJ17EDSTmpANjbByxbld8YKNXfix6iodeql5I_VvueFFoONDZAoSrEMChjOYFYjAlb324YShYqe9QlWTwMMRiq2FnCkSCbqKTIapxWHPlNGDDuAuoTRIYSpL1ZSqqgrnbcJfQkdahI2qiAhvIgJeWTTkhSsCiBmLncvRcQGd9-anF4OeWGVxA1hBHEdN5--li4UrscJOhfshx6myNg5ClKZHqBX20O_eoBeSxnb7pqQYDd8bKQM3s2_I5BtiF04vtjWdQ9v_-TH1r0DW70OqfPNMMO_QUS7lTRaDI-yVadHK6NqhSKLm-MEkqQYtN0eZIWVWxuMDCPglZRHqoQq5oXYM6BEJT-dn99rCzg2s7xH2AZqgP8bvMco_OtIx1pTsYC6ebIT-i42vdF30fLGAgu-2vuL8YnmEeNABONlpTZvjcpqz04JjUf3CfyaHYw9qcWJ-4nQoljl7ATZ7QTWYXnuOKU8vIcX6FZhqZfYyDj2lXd0yudDx-bze8dGItXgw2704Vyjd0ZSOJORadO5LxtkP61AssMACGU-b1dw07CT1JZTeIPlPIq-K6Wxiw9cH3_RGWq0adwU48Fw791Pg1PNvCh_gx34bCzfXqCl9VahsIorcHQcsf8tnjGFW6WSn_5zHaXwsOx-RbysmzNguuwPfmwwfA-7PUNo_ztGri0u5ADmLA6m5vdqmaEIuLBx6G_3qA17sOZd2_YTDmdulJ5Y_EhszIgBaRUF_T8J6EdgUXLXmiETLv-tlIW_slFvW_IBSWty8l5QEWG5YWOThnjX1smPaillXOj7_ISQySWWA0bwMOhb2U7lcdgIbKZjgjBdnipjeoI9FrZPIQMibHlw8GWrSrVuATaKR5CGdTYaLHWkHJjNiJ9GvXkud)](https://www.plantuml.com/plantuml/uml/pLNVQzim47xNNt4pAqqXSTgGXn4ejEHfi4BPElOmob1i9T4gMqQIwxKb__i-HJPHLo-bbFsu1rUwq-bEttTVSsOTDwuk5EDAwQQGFsIvOkoGHgCHleSbByICvbEggqgOZ5j10oLl1BcreALMkJ17EDSTmpANjbByxbld8YKNXfix6iodeql5I_VvueFFoONDZAoSrEMChjOYFYjAlb324YShYqe9QlWTwMMRiq2FnCkSCbqKTIapxWHPlNGDDuAuoTRIYSpL1ZSqqgrnbcJfQkdahI2qiAhvIgJeWTTkhSsCiBmLncvRcQGd9-anF4OeWGVxA1hBHEdN5--li4UrscJOhfshx6myNg5ClKZHqBX20O_eoBeSxnb7pqQYDd8bKQM3s2_I5BtiF04vtjWdQ9v_-TH1r0DW70OqfPNMMO_QUS7lTRaDI-yVadHK6NqhSKLm-MEkqQYtN0eZIWVWxuMDCPglZRHqoQq5oXYM6BEJT-dn99rCzg2s7xH2AZqgP8bvMco_OtIx1pTsYC6ebIT-i42vdF30fLGAgu-2vuL8YnmEeNABONlpTZvjcpqz04JjUf3CfyaHYw9qcWJ-4nQoljl7ATZ7QTWYXnuOKU8vIcX6FZhqZfYyDj2lXd0yudDx-bze8dGItXgw2704Vyjd0ZSOJORadO5LxtkP61AssMACGU-b1dw07CT1JZTeIPlPIq-K6Wxiw9cH3_RGWq0adwU48Fw791Pg1PNvCh_gx34bCzfXqCl9VahsIorcHQcsf8tnjGFW6WSn_5zHaXwsOx-RbysmzNguuwPfmwwfA-7PUNo_ztGri0u5ADmLA6m5vdqmaEIuLBx6G_3qA17sOZd2_YTDmdulJ5Y_EhszIgBaRUF_T8J6EdgUXLXmiETLv-tlIW_slFvW_IBSWty8l5QEWG5YWOThnjX1smPaillXOj7_ISQySWWA0bwMOhb2U7lcdgIbKZjgjBdnipjeoI9FrZPIQMibHlw8GWrSrVuATaKR5CGdTYaLHWkHJjNiJ9GvXkud)

**Reply by The-Lum** [reply]
Hello @kirchsth,

Yes it is plan...

Here is the imagined roadmap:
1. Creation of `readme.md` with YAML frontmatter (✔️with last PR)
2. Suppress `INFO` file et use of the new `readme.md` 🔜 🆕 
3. Update & create new getter to return those values directly on plantuml ⏭️ 
4. Then create new built-in function to use those value directly on the preproceesor ⏭️ 


Awaiting all those enhancements... And in the meantime you can always use functions like:
```puml
@startuml
!$stdlib=%get_all_stdlib(detailed)
!function $libversion($lib)
!return $stdlib[$lib][version]
!end function

file f [
Version of C4: $libversion(c4)
]
@enduml
```
>[![](https://www.planttext.com/api/plantuml/svg/LOv12iCW44NtWdUOGWhfFgj28T0xT2CXs3WMmHhGCUVl5zj5TpFp_d_CNCHcgU-erMaeuc9utiulbjN6kFPzT2msH7ONP7nDcuGzqG1oS2wOnpOtc5bg1kezqywB-OOMS4wEVWAjjFAmaYUZrQE7QFTqdwv_ycs26kqPRVpv0G00)](https://www.planttext.com/?text=LOv12iCW44NtWdUOGWhfFgj28T0xT2CXs3WMmHhGCUVl5zj5TpFp_d_CNCHcgU-erMaeuc9utiulbjN6kFPzT2msH7ONP7nDcuGzqG1oS2wOnpOtc5bg1kezqywB-OOMS4wEVWAjjFAmaYUZrQE7QFTqdwv_ycs26kqPRVpv0G00) 

**_Enjoy,_**
Regards,
Th.

**Comment by arnaudroques** [comment]
Yes, that's what we planned to do.

The only challenge is to find the right syntax.

And here, you can help us :-)

Here are a few proposals:

- `%stdlibinfo()`: Returns all information about all libraries in JSON format.  
- `%stdlibinfo("C4")`: Returns all information about the library "C4" in JSON format.  
- `%stdlibinfo("C4", "description")`: Returns only the "description" field.  
- `%stdlib()`: Returns all information about all libraries in JSON format.  
- `%stdlib("C4")`: Returns all information about the library "C4" in JSON format.  
- `%stdlib("C4", "description")`: Returns only the "description" field.  

Your feedback is welcome!

**Comment by kirchsth** [comment]
Hi @The-Lum, @arnaudroques 

I didn't know that PlantUML has already the `%get_all_stdlib()` function therefore I suggested a similar functionality.
Maybe the simplest is to add a `%get_all_stdlib(full)` that it returns all new properties too. 

And  optional add something like 
   - `%get_stdlib(name)` -> returns version number of the concrete standard library
   - `%get_stdlib(name, detailed)` -> returns name, version, source 
   - `%get_stdlib(name, full)` -> returns all properties

If "c4" is used as stdlib-name (and not the upper case "C4", which is the folder name) we should add a **stdlib-specific** documentation link, and the documentation contains the correct folder name "C4". Eg.  https://github.com/plantuml/plantuml-stdlib?tab=readme-ov-file#c4-library-c4-plantuml-c4 (and of course the C4 documentation itself has to be updated too ).

Another option could be that the [plantuml stdlib Readme.md](https://github.com/plantuml/plantuml-stdlib/blob/master/README.md) contains only an overview (table?) of all stdlibs with links to the stdlib specific readme.md in the subfolders.
 
Best regards
Helmut

PS.: related to documentation: ex. `%get_all_stdlib()`, `%get_all_theme()` and `%random()` is not part of the [Buildin functions](https://plantuml.com/preprocessing#291cabbe982ff775), maybe you can update it that all functions are listed in the grid.
Additional the table could be (grouped and) sorted.

**Reply by The-Lum** [reply]
@kirchsth 

>PS.: related to documentation: ex. %get_all_stdlib(), %get_all_theme() and %random() is not part of the [Builtin functions](https://plantuml.com/preprocessing#291cabbe982ff775), maybe you can update it that all functions are listed in the grid.

Thanks for the report, the change had just been made to the doc. _(remains JSON Builtin functions)_:
- https://plantuml.com/preprocessing#291cabbe982ff775

Regards,
Th.

**Comment by arnaudroques** [comment]
With [last snapshot 1.2024.9beta1](https://github.com/plantuml/plantuml/releases/tag/snapshot), you can now have:

```
@startjson
%get_stdlib()
@endjson
```

<hr>

You can get information about a specific library.
I realize that there is a confusion between `c4` and `C4`.
So probably we should simply ignore case, so `%get_stdlib(c4)` and `%get_stdlib(C4)` would both work. This is not implemented yet.

```
@startjson
%get_stdlib(c4)
@endjson
```

<hr>

The following calls return Strings, not Json objects:

```
@startuml
Alice -> Bob: %get_stdlib(c4, foo)
Alice -> Bob: %get_stdlib(c4, version)
Alice -> Bob: %get_stdlib(c4, display_name)
@enduml
```

This is not a final version, only a working prototype :-)
Any information added into `README.md` file will be available, so you can even add new key entries there if needed.

Any though?

**Reply by kirchsth** [reply]
Hi @arnaudroques,

thank you it works perfect. I think we can do everything with it now

```plantuml
@startuml
!include <C4/C4_Container>

json get_stdlib %get_stdlib()
json get_stdlib_C4 %get_stdlib(c4)

!$data=%get_stdlib()

[a]
note right of a

get_stdlib(c4):
%get_stdlib(c4)

get_stdlib(c4, name): %get_stdlib(c4, name)
get_stdlib(c4, source): %get_stdlib(c4, source)
get_stdlib(c4, display_name): %get_stdlib(c4, display_name)
get_stdlib(c4, version): %get_stdlib(c4, version)

data[c4][version]: $data[c4][version]

end note

C4VersionDetails()
@enduml
``` 

Thank you
Helmut

**Comment by kirchsth** [comment]
Hi @arnaudroques, @The-Lum and @Potherca

I tried to create a "real" sample of a README.md based on the C4-PlantUML stdlib.

[C4-PlantUML stdlib README.md](https://github.com/kirchsth/plantuml-stdlib/tree/102-discussion-readme-c4-plantuml/stdlib/C4)

1) the table is basically only a copy of yaml (remove in the final version?)
    ? Is it possible to change the format of the yaml, that it vertical and not horizontal formatted ?

2) Remaining text is the current "C4 (C4-PlantUML)" section of the "plantuml-stdlib"-README
    ? is it possible to copy it from the "C4" README into "plantuml-stdlib"-README ?
         
Can you please review it and add your thoughts?

Thank you and best regards
Helmut

**Reply by kirchsth** [reply]
I added 2 badges based alternatives (links are missing) that the required space can be reduced:
a) first alternative contains all properties, incl. the orig. source (color orange) related properties
b) second alternative contains only the "info" of the copied source and the "orig. source properties" are only visible via the yaml-properties of the linked C4-README.

**Other topic license:**
Who checks the compatibility of the differences licenses, or can we assume that all licenses are allowed as soon the files are copied into the plantuml-stdlib (if this is the case then we can remove license entry in the second badges version too)?
